libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+
+        /**
+         * This monitors the life time of the job that handles the channels. If that job is terminated, we cleanup.
+         */
+        jobControl.scope.launch {
+            select<Unit> {
+                cleanupJob.onJoin {
+                    locationEngine.removeLocationUpdates(locationCallback)
+                }
+            }
+        }
+    }
+
+    private suspend fun getDatePartitionedLocations(predicate: (Location) -> Boolean): Array<Location> {
+        val retVal = mutableListOf<Location>()
+        channelLocationBuffer.send(LocationBufferControl(LocationBufferCommands.BUFFER_READ, null) { location ->
+            if (predicate(location)) {
+                retVal.add(location)
+            }
+        })
+        return retVal.toTypedArray()
+    }
+
+    private fun parseRouteProgressAsync(): Deferred<TelemetryStep> {
+        return jobControl.scope.async {
+            val telemetryStep = TelemetryStep() // Default initialization
+            val routeProgress = channelOnRouteProgress.receive()
+            routeProgress.currentLegProgress()?.upcomingStep()?.let { legStep ->
+                telemetryStep.distance = legStep.distance().toInt()
+                telemetryStep.distanceRemaining = routeProgress.distanceRemaining().toInt()
+                telemetryStep.duration = legStep.duration().toInt()
+                telemetryStep.durationRemaining = routeProgress.durationRemaining().toInt()
+                telemetryStep.previousModifier = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.modifier()
+                        ?: ""
+                telemetryStep.previousName = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.name()
+                        ?: ""
+                telemetryStep.previousType = routeProgress.currentLegProgress()?.currentStepProgress()?.step()?.maneuver()?.type()
+                        ?: ""
+                telemetryStep.upcomingType = legStep.maneuver().type() ?: ""
+                telemetryStep.upcomingModifier = legStep.maneuver().modifier() ?: ""
+                telemetryStep.upcomingName = legStep.name() ?: ""
+            }
+            telemetryStep
+        }
+    }
+
+    /**
+     * TODO:OZ add one of these for each telemetry event
+     * Generates a well-formed Feedback event.
+     */
+    private suspend fun populateUserFeedbackEvent(event: TelemetryEventFeedback): TelemetryEventInterface {
+        val phoneState = PhoneState(context)
+
+        val feedbackEvent = TelemetryUserFeedbackWrapper(event,
+                phoneState.userId,
+                phoneState.audioType,
+                getDatePartitionedLocations { location -> location.time < System.currentTimeMillis() },
+                getDatePartitionedLocations { location -> location.time > System.currentTimeMillis() },
+                UUID.randomUUID().toString(),
+                event.screenShot,
+                parseRouteProgressAsync().await()
```

Comment:
why do you use `async-await` here? 
Looks like `withContext()` is enough.

Reply:
Async is used, because populateUserFeedbackEvent() returns a value. Look at line #271. That is  the return value of  parseRouteProgressAsync() which becomes the return value for populateUserFeedbackEvent().

Reply:
withContext also returns a value.
if you call `async` function and `await` at the same line - you use it in a wrong way

Reply:
I believe the change you're looking for is this
private suspend fun parseRouteProgressAsync(). Then I can remove the call to async().
👍